ape_solidity/exceptions.py

```diff
@@ -7,3 +10,139 @@ def __init__(self):
+        super().__init__(RuntimeErrorType.ASSERTION_ERROR, message, **kwargs)
+
+
+class DivisionByZeroError(SolidityRuntimeError, ZeroDivisionError):
+    """
+    Raised when dividing goes wrong (such as using a 0 denominator).
+    """
+
+    def __init__(self, **kwargs):
+        message = "Division or modulo division by zero"
+        super().__init__(RuntimeErrorType.DIVISION_BY_ZERO_ERROR, message, **kwargs)
+
+
+class EnumConversionError(SolidityRuntimeError):
+    """
+    Raised when Solidity fails to convert an enum value to its primitive type.
+    """
+
+    def __init__(self, **kwargs):
+        message = "Tried to convert a value into an enum, but the value was too big or negative."
+        super().__init__(RuntimeErrorType.ENUM_CONVERSION_OUT_OF_BOUNDS, message, **kwargs)
+
+
+class EncodeStorageError(SolidityRuntimeError):
+    """
+    Raised when Solidity fails to encode a storage value.
+    """
+
+    def __init__(self, **kwargs):
+        message = "Incorrectly encoded storage byte array."
+        super().__init__(RuntimeErrorType.INCORRECTLY_ENCODED_STORAGE_BYTE_ARRAY, message, **kwargs)
+
+
+class IndexOutOfBoundsError(SolidityRuntimeError, IndexError):
+    """
+    Raised when accessing an index that is out of bounds in your contract.
+    """
+
+    def __init__(self, **kwargs):
+        message = "Array accessed at an out-of-bounds or negative index."
+        super().__init__(RuntimeErrorType.INDEX_OUT_OF_BOUNDS_ERROR, message, **kwargs)
+
+
+class MemoryOverflowError(SolidityRuntimeError, OverflowError):
+    """
+    Raised when exceeding the allocating memory for a data type
+    in Solidity.
+    """
+
+    def __init__(self, **kwargs):
+        message = "Too much memory was allocated, or an array was created that is too large."
+        super().__init__(RuntimeErrorType.MEMORY_OVERFLOW_ERROR, message, **kwargs)
+
+
+class PopOnEmptyArrayError(SolidityRuntimeError):
+    """
+    Raised when popping from a data-structure fails in your contract.
+    """
+
+    def __init__(self, **kwargs):
+        message = ".pop() was called on an empty array."
+        super().__init__(RuntimeErrorType.POP_ON_EMPTY_ARRAY, message, **kwargs)
+
+
+class ZeroInitializedVariableError(SolidityRuntimeError):
+    """
+    Raised when calling a zero-initialized variable of internal function type.
+    """
+
+    def __init__(self, **kwargs):
+        message = "Called a zero-initialized variable of internal function type."
+        super().__init__(RuntimeErrorType.ZERO_INITIALIZED_VARIABLE, message, **kwargs)
+
+
+RUNTIME_ERROR_CODE_PREFIX = "0x4e487b71"
+RuntimeErrorUnion = Union[
+    SolidityArithmeticError,
+    SolidityAssertionError,
+    DivisionByZeroError,
+    EnumConversionError,
+    EncodeStorageError,
+    IndexOutOfBoundsError,
+    MemoryOverflowError,
+    PopOnEmptyArrayError,
+    ZeroInitializedVariableError,
+]
+RUNTIME_ERROR_MAP: Dict[RuntimeErrorType, Type[RuntimeErrorUnion]] = {
```

Comment:
Can simplify by base class
```suggestion
RUNTIME_ERROR_MAP: Dict[RuntimeErrorType, Type[SolidityRuntimeError]] = {
```

Reply:
doesnt actually work with type checking because of the constructor of the base class... cant make assumptions about the constructor if you do that in the code... causing the need for a type: ignore.
this is better than type ignore